File: src/ladym/layers/semantic.py
"""L2 — Semantic memory.

Consolidated facts *and* codebase analysis live here, distinguished only by ``type``.
Writes go through the consolidation primitive (``operations/consolidate.py``) so that the
``ADD/UPDATE/DELETE/NOOP`` decision from mem0 is honoured and duplicates don't accumulate.
"""

from __future__ import annotations

import hashlib
from typing import Any

from ..schema import Layer, Memory, MemoryType
from ..storage.embeddings import EmbeddingProvider
from ..storage.store import SQLiteStore


def content_hash(text: str) -> str:
    return hashlib.blake2b(text.encode(), digest_size=16).hexdigest()


class SemanticMemory:
    def __init__(self, store: SQLiteStore, embedder: EmbeddingProvider,
                 workspace: str = "default"):
        self.store = store
        self.embedder = embedder
        self.workspace = workspace

    def put_fact(self, content: str, *, summary: str = "",
                 tags: list[str] | None = None, metadata: dict[str, Any] | None = None,
                 source: str = "") -> Memory:
        """Direct fact write. Used by consolidate() and remember()."""
        mem = Memory(
            layer=Layer.SEMANTIC,
            type=MemoryType.FACT,
            content=content,
            summary=summary or content[:80],
            tags=tags or [],
            metadata=metadata or {},
            source=source,
            workspace=self.workspace,
            content_hash=content_hash(content),
        )
        vec = self.embedder.embed(content)
        self.store.put_memory(mem, vector=vec)
        return mem

    def put_code_file(self, file_path: str, summary: str, *,
                      language: str = "") -> Memory:
        mem = Memory(
            layer=Layer.SEMANTIC,
            type=MemoryType.CODE_FILE,
            content=f"{file_path}: {summary}",
            summary=summary[:120],
            tags=["code", language] if language else ["code"],
            metadata={"file_path": file_path, "language": language},
            source=file_path,
            workspace=self.workspace,
            content_hash=content_hash(file_path + "|" + summary),
        )
        self.store.put_memory(mem, vector=self.embedder.embed(mem.content))
        return mem

    def find_by_hash(self, h: str) -> Memory | None:
        return self.store.find_by_hash(h, workspace=self.workspace)

File: layers/semantic.go
package layers

import (
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// SemanticMemory is L2 — consolidated facts and codebase analysis.
type SemanticMemory struct {
	Store     storage.Store
	Embedder  storage.EmbeddingProvider
	Workspace string
}

// NewSemanticMemory builds a SemanticMemory.
func NewSemanticMemory(store storage.Store, embedder storage.EmbeddingProvider, workspace string) *SemanticMemory {
	if workspace == "" {
		workspace = "default"
	}
	return &SemanticMemory{Store: store, Embedder: embedder, Workspace: workspace}
}

// PutFact writes a semantic fact directly.
func (s *SemanticMemory) PutFact(content, summary string, tags []string, metadata map[string]any, source string) (*schema.Memory, error) {
	if summary == "" {
		summary = truncate(content, 80)
	}
	if tags == nil {
		tags = []string{}
	}
	if metadata == nil {
		metadata = map[string]any{}
	}
	m := schema.NewMemory(schema.LayerSemantic, schema.TypeFact)
	m.Content = content
	m.Summary = summary
	m.Tags = tags
	m.Metadata = metadata
	m.Source = source
	m.Workspace = s.Workspace
	m.ContentHash = schema.ContentHash(content)

	vec, err := s.Embedder.Embed(content)
	if err != nil {
		return nil, err
	}
	if err := s.Store.PutMemory(m, vec); err != nil {
		return nil, err
	}
	return m, nil
}

// PutCodeFile writes a whole-file summary as L2 code memory.
func (s *SemanticMemory) PutCodeFile(filePath, summary, language string) (*schema.Memory, error) {
	tags := []string{"code"}
	if language != "" {
		tags = append(tags, language)
	}
	m := schema.NewMemory(schema.LayerSemantic, schema.TypeCodeFile)
	m.Content = filePath + ": " + summary
	m.Summary = truncate(summary, 120)
	m.Tags = tags
	m.Metadata = map[string]any{"file_path": filePath, "language": language}
	m.Source = filePath
	m.Workspace = s.Workspace
	m.ContentHash = schema.ContentHash(filePath + "|" + summary)

	vec, err := s.Embedder.Embed(m.Content)
	if err != nil {
		return nil, err
	}
	if err := s.Store.PutMemory(m, vec); err != nil {
		return nil, err
	}
	return m, nil
}

// FindByHash returns the memory with the given content hash.
func (s *SemanticMemory) FindByHash(h string) (*schema.Memory, error) {
	return s.Store.FindByHash(h, s.Workspace)
}
